### PR TITLE
[Merged by Bors] - refactor: use `NeZero` for measures

### PR DESCRIPTION
**Pull request by urkud**
Assume `NeZero μ` instead of `μ.ae.NeBot` everywhere,
and sometimes instead of `μ ≠ 0`.

## API changes

- `Convex.average_mem`, `Convex.set_average_mem`,
  `ConvexOn.average_mem_epigraph`, `ConcaveOn.average_mem_hypograph`,
  `ConvexOn.map_average_le`, `ConcaveOn.le_map_average`:
  assume `[NeZero μ]` instead of `μ ≠ 0`;
- `MeasureTheory.condexp_bot'`, `essSup_const'`, `essInf_const'`,
  `MeasureTheory.laverage_const`, `MeasureTheory.laverage_one`,
  `MeasureTheory.average_const`:
  assume `[NeZero μ]` instead of `[μ.ae.NeBot]`
- `MeasureTheory.Measure.measure_ne_zero`: replace with an instance;
- remove `@[simp]` from `MeasureTheory.ae_restrict_neBot`,
  use `≠ 0` in the RHS;
- turn `MeasureTheory.IsProbabilityMeasure.ae_neBot`
  into a theorem because `inferInstance` can find it now;
- add instances:
  - `[NeZero μ] : NeZero (μ univ)`;
  - `[NeZero (μ s)] : NeZero (μ.restrict s)`;
  - `[NeZero μ] : μ.ae.NeBot`;
  - `[IsProbabilityMeasure μ] : NeZero μ`;
  - `[IsFiniteMeasure μ] [NeZero μ] : IsProbabilityMeasure ((μ univ)⁻¹ • μ)`
    this was a theorem `MeasureTheory.isProbabilityMeasureSmul` assuming `μ ≠ 0`;

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ocfnash** [comment]
We should also apply this change to [[MeasureTheory.Measure.measure_ne_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/OpenPos.html#MeasureTheory.Measure.measure_ne_zero)](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/OpenPos.html#MeasureTheory.Measure.measure_ne_zero) right?

**Comment by urkud** [comment]
> We should also apply this change to [MeasureTheory.Measure.measure_ne_zero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Measure/OpenPos.html#MeasureTheory.Measure.measure_ne_zero) right?

Done.

File: Mathlib/MeasureTheory/Function/LpSeminorm.lean [hunk 1]
```diff
@@ -437,14 +437,13 @@ theorem snormEssSup_lt_top_of_ae_bound {f : α → F} {C : ℝ} (hfC : ∀ᵐ x
...
-  · simp [hμ]
-  haveI : μ.ae.NeBot := ae_neBot.mpr hμ
+  rcases eq_zero_or_neZero μ with rfl | hμ
+  · simp
   by_cases hp : p = 0
   · simp [hp]
   have : ∀ᵐ x ∂μ, ‖f x‖₊ ≤ ‖(C : ℝ)‖₊ := hfC.mono fun x hx => hx.trans_eq C.nnnorm_eq.symm
   refine' (snorm_mono_ae this).trans_eq _
-  rw [snorm_const _ hp hμ, C.nnnorm_eq, one_div, ENNReal.smul_def, smul_eq_mul]
+  rw [snorm_const _ hp (NeZero.ne μ), C.nnnorm_eq, one_div, ENNReal.smul_def, smul_eq_mul]
```
Comment (ocfnash) [inline Mathlib/MeasureTheory/Function/LpSeminorm.lean:446]: Should `snorm_const` also be changed to consume `[NeZero μ]` rather than `(hμ : μ ≠ 0)`?

**Review by ocfnash — APPROVED** [review]
Thanks for doing this.

I think there is room for a second refactor dealing with lemmas like `snorm_const` which could have an argument of the form `(hμ : μ ≠ 0)` turned into `[NeZero μ]`. Indeed if I search for the regex `(ν|μ) ≠ 0` I find 61 occurrences and I expect most / all would be better using `NeZero`.

However I think this can happen separately: the changes here already constitute a strict improvement on what we have.

bors merge

**Pull request closed without merging**